File: go/src/kanzi/bitstream/DebugInputBitStream.go
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package bitstream

import (
	"errors"
	"fmt"
	"io"
	"kanzi"
)

type DebugInputBitStream struct {
	delegate  kanzi.InputBitStream
	out       io.Writer
	mark      bool
	hexa      bool
	current   byte
	width     int
	lineIndex int
}

func NewDebugInputBitStream(ibs kanzi.InputBitStream, writer io.Writer) (*DebugInputBitStream, error) {
	if ibs == nil {
		return nil, errors.New("The delegate cannot be null")
	}

	if writer == nil {
		return nil, errors.New("The writer cannot be null")
	}

	this := new(DebugInputBitStream)
	this.delegate = ibs
	this.out = writer
	this.width = 80
	return this, nil
}

func (this *DebugInputBitStream) ReadBit() int {
	res := this.delegate.ReadBit()

	this.current <<= 1
	this.current |= byte(res)
	this.lineIndex++

	if res&1 == 1 {
		fmt.Fprintf(this.out, "1")
	} else {
		fmt.Fprintf(this.out, "0")
	}

	if this.mark == true {
		fmt.Fprintf(this.out, "r")
	}

	if this.width > 7 {
		if (this.lineIndex-1)%this.width == this.width-1 {
			if this.hexa == true {
				fmt.Fprintf(this.out, "[%d] ", this.current)
			}

			fmt.Fprintf(this.out, "\n")
			this.lineIndex = 0
		} else if this.lineIndex&7 == 0 {
			fmt.Fprintf(this.out, " ")

			if this.hexa == true {
				fmt.Fprintf(this.out, "[%d] ", this.current)
			}
		}
	} else if this.lineIndex&7 == 0 {
		fmt.Fprintf(this.out, " ")

		if this.hexa == true {
			fmt.Fprintf(this.out, "[%d] ", this.current)
		}
	}

	return res
}

func (this *DebugInputBitStream) ReadBits(length uint) uint64 {
	res := this.delegate.ReadBits(length)

	for i := uint(1); i <= length; i++ {
		bit := (res >> (length - i)) & 1
		this.current <<= 1
		this.current |= byte(bit)
		this.lineIndex++
		fmt.Fprintf(this.out, "%d", bit)

		if this.mark == true && i == length {
			fmt.Fprintf(this.out, "r")
		}

		if this.width > 7 {
			if this.lineIndex%this.width == 0 {
				if this.hexa == true {
					fmt.Fprintf(this.out, "[%d] ", this.current)
				}

				fmt.Fprintf(this.out, "\n")
				this.lineIndex = 0
			} else if this.lineIndex&7 == 0 {
				fmt.Fprintf(this.out, " ")

				if this.hexa == true {
					fmt.Fprintf(this.out, "[%d] ", this.current)
				}
			}
		} else if this.lineIndex&7 == 0 {
			fmt.Fprintf(this.out, " ")

			if this.hexa == true {
				fmt.Fprintf(this.out, "[%d] ", this.current)
			}
		}
	}

	return res
}

func (this *DebugInputBitStream) HasMoreToRead() (bool, error) {
	return this.delegate.HasMoreToRead()
}

func (this *DebugInputBitStream) Close() (bool, error) {
	return this.delegate.Close()
}

func (this *DebugInputBitStream) Read() uint64 {
	return this.delegate.Read()
}

func (this *DebugInputBitStream) Mark(mark bool) {
	this.mark = mark
}

func (this *DebugInputBitStream) ShowByte(show bool) {
	this.hexa = show
}

File: java/src/main/java/kanzi/bitstream/DebugInputBitStream.java
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.bitstream;

import kanzi.BitStreamException;
import kanzi.InputBitStream;
import java.io.PrintStream;


// Very util little wrapper used to print the bits written to the delegate
// bitstream (uses the decorator design pattern)
public final class DebugInputBitStream implements InputBitStream
{
    private final InputBitStream delegate;
    private final PrintStream out;
    private final int width;
    private int lineIndex;
    private boolean mark;
    private boolean hexa;
    private byte current;
    

    public DebugInputBitStream(InputBitStream bitstream)
    {
        this(bitstream, System.out, 80);
    }


    public DebugInputBitStream(InputBitStream bitstream, PrintStream out)
    {
        this(bitstream, out, 80);
    }


    public DebugInputBitStream(InputBitStream bitstream, PrintStream out, int width)
    {
        if (bitstream == null)
            throw new NullPointerException("Invalid null bitstream parameter");
        
        if (out == null)
            throw new NullPointerException("Invalid null print stream parameter");
        
        if ((width != -1) && (width < 8))
            width = 8;
        
        if (width != -1)
            width &= 0xFFFFFFF8;
        
        this.width = width;
        this.delegate = bitstream;
        this.out = out;
    }
    
    
    public synchronized void setMark(boolean mark)
    {
        this.mark = mark;
    }
    
    
    public synchronized boolean mark()
    {
        return this.mark;
    }
    
    
    public synchronized void showByte(boolean hex)
    {
        this.hexa = hex;
    }
    
    
    public synchronized boolean showByte()
    {
        return this.hexa;
    }
    
    
    protected synchronized void printByte(byte val)
    {
       if ((val >= 0) && (val < 10))
            this.out.print(" [00" + (val & 0xFF) + "] ");
        else if ((val >= 0) && (val < 100))
            this.out.print(" [0" + (val & 0xFF) + "] ");
        else
            this.out.print(" [" + (val & 0xFF) + "] ");
    }
    
    
    // Returns 1 or 0
    @Override
    public synchronized int readBit() throws BitStreamException
    {
         int res = this.delegate.readBit();
         this.current <<= 1;
         this.current |= res;
         this.out.print((res & 1) == 1 ? "1" : "0");
         this.lineIndex++;

         if (this.mark == true)
             this.out.print("r");

         if (this.width != -1)
         {
             if ((this.lineIndex-1) % this.width == this.width-1)
             {
                 if (this.showByte())
                     this.printByte(this.current);

                 this.out.println();
                 this.lineIndex = 0;
             }
             else if ((this.lineIndex & 7) == 0)
             {
                 if (this.showByte())
                     this.printByte(this.current);
                else
                    this.out.print(" "); 
             }
         }
         else if ((this.lineIndex & 7) == 0)
         {
             if (this.showByte())
                 this.printByte(this.current);
                else
                    this.out.print(" "); 
         }
        
        return res;
    }
    
    
    @Override
    public synchronized long readBits(int length) throws BitStreamException
    {
        long res = this.delegate.readBits(length);
        
        for (int i=1; i<=length; i++)
        {
            long bit = (res >> (length-i)) & 1;
            this.lineIndex++;
            this.current <<= 1;
            this.current |= bit;
            this.out.print((bit == 1) ? "1" : "0");
            
            if ((this.mark == true) && (i == length))
                this.out.print("r");
            
            if (this.width != -1)
            {
                if (this.lineIndex % this.width == 0)
                {
                    if (this.showByte())
                        this.printByte(this.current);
                    
                    this.out.println();
                    this.lineIndex = 0;
                }
                else if ((this.lineIndex & 7) == 0)
                {
                    if (this.showByte())
                        this.printByte(this.current);
                    else
                        this.out.print(" "); 
                }
            }
            else if ((this.lineIndex & 7) == 0)
            {
                if (this.showByte())
                    this.printByte(this.current);
                else
                    this.out.print(" "); 
            }
        }
        
        return res;
    }
        
    
    @Override
    public void close() throws BitStreamException
    {
        this.delegate.close();
    }
    
    
    @Override
    public long read()
    {
        return this.delegate.read();
    }
    
    
    @Override
    public boolean hasMoreToRead()    
    {
        return this.delegate.hasMoreToRead();
    }
}
